calico/felix/selectors.py

```diff
@@ -0,0 +1,664 @@
+    """Parse action for an 'in' AST node."""
+    lhs, rhs = tokens
+    assert isinstance(lhs, Label)
+    assert isinstance(rhs, SetLiteral)
+    return LabelInSetLiteral(lhs.label_name, rhs.value)
+
+
+def not_in_op(parse_str=None, location=None, tokens=None):
+    """Parse action for a 'not in' AST node."""
+    lhs, rhs = tokens
+    return NotInOp(lhs, rhs)
+
+
+class BinaryOp(ExprNode):
+    """
+    Base class for binary operators.  Provides common function, such as
+    generating a hash code and simple evaluation.
+    """
+    __slots__ = ["lhs", "rhs"]
+    operation = None
+    operation_str = None
+
+    def __init__(self, lhs, rhs):
+        self.lhs = lhs
+        self.rhs = rhs
+
+    def evaluate(self, labels):
+        return self.operation(self.lhs.evaluate(labels),
+                              self.rhs.evaluate(labels))
+
+    def __hash__(self):
+        h = hash(self.__class__)
+        h = h * 37 + hash(self.lhs)
+        h = h * 37 + hash(self.rhs)
+        return h
+
+    def __eq__(self, other):
+        return (type(other) == type(self) and
+                self.lhs == other.lhs and
+                self.rhs == other.rhs)
+
+    def collect_str_fragments(self, fragment_list):
+        self.lhs.collect_str_fragments(fragment_list)
+        fragment_list.append(" ")
+        fragment_list.append(self.operation_str)
+        fragment_list.append(" ")
+        self.rhs.collect_str_fragments(fragment_list)
+
+
+class LabelToLiteralEquality(BinaryOp):
+    """
+    Represents the sub-expression label == "value".
+
+    The LHS stores the name of the label directly, the RHS stores the
+    value of the string literal.
+    """
+    __slots__ = []
+    operation = operator.eq
+    operation_str = "=="
+
+    def evaluate(self, labels):
+        # Since we store the label name and value directly, we need a
+        # customized eval function...
+        return labels.get(self.lhs) == self.rhs
+
+    def collect_reqd_values(self, pr_set):
+        pr_set.add((self.lhs, self.rhs))
+
+    def collect_str_fragments(self, fragment_list):
+        fragment_list.append(self.lhs)
+        fragment_list.append(" == ")
+        fragment_list.append(repr(self.rhs))
+
+
+class LabelInSetLiteral(BinaryOp):
+    """
+    Represents the sub-expression label in {"value", "value2" ...}.
+
+    The LHS stores the name of the label directly, the RHS stores the
+    value of the set literal.
+    """
+    __slots__ = []
+    operation_str = "in"
+
+    def evaluate(self, labels):
+        # Since we store the label name and value directly, we need a
+        # customized eval function...
+        return labels.get(self.lhs) in self.rhs
+
+    def collect_reqd_values(self, pr_set):
+        if len(self.rhs) == 1:
+            # We don't have a way to represent alternatives yet so we can only
+            # express a requirement if there's only one entry in the set.
+            pr_set.update(self.rhs)
```

Comment:
As discussed, returning an empty set is a safe approximation.  It'll result in falling back on other requirements found in the expression or just failing to optimise the expression.